lib/components/primitive-components/Trace/Trace.ts

```diff
@@ -596,6 +596,157 @@ export class Trace
+          },
+          to: {
+            x: fromAnchorPos.x,
+            y: fromAnchorPos.y + fromHandelUp * fromSchBoxVertOffset,
+          },
+        },
+
+        {
+          from: {
+            x: fromAnchorPos.x,
+            y: fromAnchorPos.y,
+          },
+          to: {
+            x: fromAnchorPos.x + fromHorzPortDirection * fromSchBoxHorzOffset,
+            y:
+              fromAnchorPos.y +
+              fromVertPortDirectionOffset * fromSchBoxVertOffset,
+          },
+        },
+      ],
+      junctions: [
+        {
+          x: fromAnchorPos.x + fromHorzPortDirection * fromSchBoxHorzOffset,
+          y:
+            fromAnchorPos.y +
+            fromVertPortDirectionOffset * fromSchBoxVertOffset,
+        },
+      ],
+      source_trace_id: this.source_trace_id!,
+    })
+
+    const toSchBoxVertOffset = toPort.parent?.config.shouldRenderAsSchematicBox
+      ? 3
+      : 1
+    const toSchBoxHorzOffset =
+      toPort.parent?.config.shouldRenderAsSchematicBox &&
+      (toPort.facingDirection === "left" || toPort.facingDirection === "right")
+        ? 3
+        : 1
+    const toHorzPortDirectionOffset =
+      toPort.facingDirection === "left"
+        ? -0.5
+        : toPort.facingDirection === "down"
+          ? 0
+          : 0.5
+    const toVertPortDirectionOffset =
+      toPort.facingDirection === "up"
+        ? 0.5
+        : toPort.facingDirection === "down"
+          ? -0.5
+          : 0
+
+    db.schematic_net_label.insert({
+      anchor_side: "top",
+      center: {
+        x: toAnchorPos.x + toHorzPortDirectionOffset * toSchBoxHorzOffset,
+        y: toAnchorPos.y + toVertPortDirectionOffset * toSchBoxVertOffset,
+      },
+      source_net_id: toPort.source_port_id!,
+      text: this.props.schDisplayLabel!,
+      anchor_position: {
+        x: toAnchorPos.x + toHorzPortDirectionOffset * toSchBoxHorzOffset,
+        y: toAnchorPos.y + toVertPortDirectionOffset * toSchBoxVertOffset,
+      },
+      symbol_name: "ground_horz",
+    })
+    const toHandelUp = toPort.facingDirection === "up" ? 0.5 : 0
+    db.schematic_trace.insert({
+      edges: [
+        {
+          from: {
+            x: toAnchorPos.x,
+            y: toAnchorPos.y,
+          },
+          to: {
+            x: toAnchorPos.x,
+            y: toAnchorPos.y + toHandelUp * toSchBoxVertOffset,
+          },
+        },
+
+        {
+          from: {
+            x: toAnchorPos.x,
+            y: toAnchorPos.y,
+          },
+          to: {
+            x: toAnchorPos.x + toHorzPortDirectionOffset * toSchBoxHorzOffset,
+            y: toAnchorPos.y + toVertPortDirectionOffset * toSchBoxVertOffset,
+          },
+        },
+      ],
+      junctions: [
+        {
+          x: toAnchorPos.x + toHorzPortDirectionOffset * toSchBoxHorzOffset,
+          y: toAnchorPos.y + toVertPortDirectionOffset * toSchBoxVertOffset,
+        },
+      ],
+      source_trace_id: this.source_trace_id!,
+    })
+  }
```

Comment:
this handling has a lot of math (the math looks pretty good) but I wonder if we could try to move it out of `Trace.ts`, possibly turning `Trace.ts` into a directory `Trace/Trace.ts` and `Trace/createDownwardNetLabelSymbol.ts`

You can use the "context passing pattern" to keep the function basically identical:

```tsx
/**
 * Write a good doc string, it's not obvious at a glance what fromPortAnchor or toPortAnchor is!
 **/
export const createDownwardNetLabelSymbol({ fromPort, toPort, fromPortAnchor, toPortAnchor }, {
 db
})
```

When you use this type of signature, you can do...

`createDownwardSymbolNetLabelSymbol({ toPort, fromPort, ... }, this)`

NOTE: We never use more than two parameters in a function, whenever there are more than two parameters we either convert to "named parameters" and take a single parameter as an object, or we switch to the context passing pattern.

When using the context-passing pattern, the most important thing is to make sure you're not inventing too many contexts. Having "{ db }" as the context is very standard/normal so it's an easy decision in this case!